Add basics of collapsible appbar

joebochill opened the pull request:
<!-- If this pull request fixes an Issue, link it below. If not, you can remove the line below -->
For PXBLUE-2111

<!-- Include a bulleted list summarizing the main changes you have made in this PR -->
#### Changes proposed in this Pull Request:
- Add AppBar component that responds to the window scroll position
- Add docs
- Add tests

<!-- Include screenshots if they will help illustrate the changes in this PR -->
#### Screenshots:
![image](https://user-images.githubusercontent.com/29152776/120349578-38faa780-c2cc-11eb-87b7-401631fbe380.png)


#### TO TEST
- In the showcase, `yarn add @n8tb1t/use-scroll-position`
- In the SharedAppBar file in the showcase, switch the `AppBar` import from @material-ui to @pxblue.

jeffvg commented:
I changed the import for AppBar but getting ./node_modules/@pxblue/react-components/core/AppBar/AppBar.js
Module not found: Can't resolve '@n8tb1t/use-scroll-position'

joebochill commented:
> I changed the import for AppBar but getting ./node_modules/@pxblue/react-components/core/AppBar/AppBar.js
> Module not found: Can't resolve '@n8tb1t/use-scroll-position'

Updated the testing instructions...you'll need to manually add that dependency in the showcase (the linking script doesn't install newly added dependencies of the library unless we have a beta out already).

jeffvg commented:
works better with FireFox browser but FireFox displays this soon as the page is scrolled...

This site appears to use a scroll-linked positioning effect. This may not work well with asynchronous panning; see https://developer.mozilla.org/docs/Mozilla/Performance/ScrollLinkedEffects for further details and to join the discussion on related tools and features!

joebochill commented:
Interesting find. I did a little reading and it looks like CSS is getting improved to handle more scroll-related behaviors to offload them away from the main javascript thread, which is cool. It looks like they don't have any properties for continuous scroll effects though (yet), which is what we are doing for our toolbar. We can keep an eye on the performance and see what we can do about throttling if it's an issue.

![image](https://user-images.githubusercontent.com/29152776/120490815-b97dde80-c386-11eb-9b90-6e04cd5b43fe.png)

jeffvg reviewed (approved):
all seems good to me except the slight janky appbar. Looks good on responsive too. Also looks like the change log needs updated

joebochill commented:
@jeffvg were you seeing jankiness even with the dev tools closed?

jeffvg commented:
> @jeffvg were you seeing jankiness even with the dev tools closed?

Yeah, the title, dropdown and toggle icons bounce as the page is scrolled

jeffvg commented:
https://user-images.githubusercontent.com/37150565/120515063-eee0f700-c39b-11eb-9321-06530652f341.mov

emclaug2 commented:
@jeffvg what browser is this on, is this on windows?  I'm not seeing any jankiness on my chrome/windows setup.

emclaug2 commented:
Firefox for me looks a bit weird; the collapsed height not being respected.  

![image](https://user-images.githubusercontent.com/6538289/120515956-dc1af200-c39c-11eb-9d00-5adc36b58c7f.png)

jeffvg commented:
> Firefox for me looks a bit weird; the collapsed height not being respected.
> 
> ![image](https://user-images.githubusercontent.com/6538289/120515956-dc1af200-c39c-11eb-9d00-5adc36b58c7f.pn

> @jeffvg what browser is this on, is this on windows? I'm not seeing any jankiness on my chrome/windows setup.

this is chrome on mac. My firefox is smooth

jeffvg commented:
> Firefox for me looks a bit weird; the collapsed height not being respected.
> 
> ![image](https://user-images.githubusercontent.com/6538289/120515956-dc1af200-c39c-11eb-9d00-5adc36b58c7f.png)

seems ok on mac with
![image](https://user-images.githubusercontent.com/37150565/120516915-e12c7100-c39d-11eb-86e1-62027bd506c0.png)

jeffvg commented:
> @jeffvg what browser is this on, is this on windows? I'm not seeing any jankiness on my chrome/windows setup.

guess I commented in wrong section. This is chrome on mac. Version 91.X

joebochill commented:
I'm aware of the jitteriness...it's definitely worse if your processor is heavily loaded. Looking for ways to improve that. The size issue should not happen, because there is a clamp on the minimum...did you change the collapsedHeightProp to something?

emclaug2 commented:
I didn't change any props; this is what I'm seeing on a firefox windows machine.  The min height is a bit weird.  Jeff can you try this on a windows machine using firefox?

joebochill commented:
> I didn't change any props; this is what I'm seeing on a firefox windows machine. The min height is a bit weird. Jeff can you try this on a windows machine using firefox?

What does it show in the inspector? Is it 64px and offscreen, or is it actually less than 64px (sorry, don't have PC to check).

emclaug2 commented:
> > I didn't change any props; this is what I'm seeing on a firefox windows machine. The min height is a bit weird. Jeff can you try this on a windows machine using firefox?
> 
> What does it show in the inspector? Is it 64px and offscreen, or is it actually less than 64px (sorry, don't have PC to check).

The collapsed height is 64px, but on my firefox it needed to be 96px for the content to fit.  That's when I released my default font-size for firefox was 24px instead of 16px, probably leftover from when we did the system font size component support.  It works as expected now!

joebochill commented:
Ah that makes sense...It's configured in px right now, while our other components are rem, hence the mismatch. Should probably switch that up.

emclaug2 commented:
> Ah that makes sense...It's configured in px right now, while our other components are rem, hence the mismatch. Should probably switch that up.

Sounds like a future enhancement 👀

huayunh commented on components/src/core/AppBar/AppBar.tsx:
```diff
@@ -0,0 +1,193 @@
+            backgroundSize: 'cover',
+            height: '100%',
+            opacity: (props: AppBarProps): number => props.backgroundImageOpacity || 0.3,
+            backgroundPosition: 'center bottom',
```
why center bottom?

huayunh commented on components/src/core/AppBar/AppBar.tsx:
```diff
@@ -0,0 +1,193 @@
+        },
+        background: {
+            position: 'absolute',
+            zIndex: -1,
```
would it make more sense to do `theme.zIndex.appBar-1`?

huayunh reviewed (commented):
Adding a gif in Readme would make more sense (in addition to the two static images you have)

joebochill commented:
Closing this for now until I make the updates to use the animation instead of 1-1
